File: 3553.minimum-weighted-subgraph-with-the-required-paths-ii.py
#
# @lc app=leetcode id=3553 lang=python3
#
# [3553] Minimum Weighted Subgraph With the Required Paths II
#

# @lc code=start
class Solution:
    def minimumWeight(self, edges: List[List[int]], queries: List[List[int]]) -> List[int]:
        from collections import defaultdict
        import sys
        sys.setrecursionlimit(1 << 20)
        n = max(max(u, v) for u, v, w in edges) + 1
        tree = defaultdict(list)
        for u, v, w in edges:
            tree[u].append((v, w))
            tree[v].append((u, w))

        LOG = 17  # 2^17 > 1e5
        parent = [[-1] * n for _ in range(LOG)]
        depth = [0] * n
        path_sum = [0] * n

        def dfs(u, p):
            for v, w in tree[u]:
                if v == p:
                    continue
                parent[0][v] = u
                depth[v] = depth[u] + 1
                path_sum[v] = path_sum[u] + w
                dfs(v, u)

        dfs(0, -1)
        for k in range(1, LOG):
            for v in range(n):
                if parent[k-1][v] != -1:
                    parent[k][v] = parent[k-1][parent[k-1][v]]

        def lca(u, v):
            if depth[u] < depth[v]:
                u, v = v, u
            for k in range(LOG-1, -1, -1):
                if parent[k][u] != -1 and depth[parent[k][u]] >= depth[v]:
                    u = parent[k][u]
            if u == v:
                return u
            for k in range(LOG-1, -1, -1):
                if parent[k][u] != -1 and parent[k][u] != parent[k][v]:
                    u = parent[k][u]
                    v = parent[k][v]
            return parent[0][u]

        def path(u, v):
            l = lca(u, v)
            return path_sum[u] + path_sum[v] - 2 * path_sum[l]

        res = []
        for src1, src2, dest in queries:
            l1 = lca(src1, dest)
            l2 = lca(src2, dest)
            l3 = lca(src1, src2)
            total = path(src1, l3) + path(src2, l3) + path(dest, l3)
            res.append(total)
        return res
# @lc code=end

File: 3553.minimum-weighted-subgraph-with-the-required-paths-ii.java
#
# @lc app=leetcode id=3553 lang=java
#
# [3553] Minimum Weighted Subgraph With the Required Paths II
#

# @lc code=start
import java.util.*;

class Solution {
    static final int LOG = 20;
    int n;
    List<int[]>[] tree;
    int[][] parent;
    int[] depth;
    long[] prefix;
    
    public int[] minimumWeight(int[][] edges, int[][] queries) {
        n = edges.length + 1;
        tree = new ArrayList[n];
        for (int i = 0; i < n; ++i) tree[i] = new ArrayList<>();
        for (int[] e : edges) {
            int u = e[0], v = e[1], w = e[2];
            tree[u].add(new int[]{v, w});
            tree[v].add(new int[]{u, w});
        }
        parent = new int[n][LOG];
        depth = new int[n];
        prefix = new long[n];
        dfs(0, -1, 0, 0);
        for (int k = 1; k < LOG; ++k) {
            for (int v = 0; v < n; ++v) {
                if (parent[v][k-1] != -1)
                    parent[v][k] = parent[parent[v][k-1]][k-1];
                else
                    parent[v][k] = -1;
            }
        }
        int[] answer = new int[queries.length];
        for (int i = 0; i < queries.length; ++i) {
            int u = queries[i][0], v = queries[i][1], d = queries[i][2];
            int lca_uv = lca(u, v);
            long sum = dist(u, d) + dist(v, d) - dist(lca_uv, d);
            answer[i] = (int)sum;
        }
        return answer;
    }
    
    void dfs(int u, int p, int d, long wsum) {
        parent[u][0] = p;
        depth[u] = d;
        prefix[u] = wsum;
        for (int[] nbr : tree[u]) {
            int v = nbr[0], w = nbr[1];
            if (v != p) {
                dfs(v, u, d+1, wsum + w);
            }
        }
    }
    
    int lca(int u, int v) {
        if (depth[u] < depth[v]) { int tmp = u; u = v; v = tmp; }
        for (int k = LOG-1; k >= 0; --k) {
            if (parent[u][k] != -1 && depth[parent[u][k]] >= depth[v])
                u = parent[u][k];
        }
        if (u == v) return u;
        for (int k = LOG-1; k >= 0; --k) {
            if (parent[u][k] != -1 && parent[u][k] != parent[v][k]) {
                u = parent[u][k];
                v = parent[v][k];
            }
        }
        return parent[u][0];
    }
    
    long dist(int u, int v) {
        int anc = lca(u, v);
        return prefix[u] + prefix[v] - 2 * prefix[anc];
    }
}
# @lc code=end